### PR TITLE
Clear jni ids on anonClass unload

### PR DESCRIPTION
**Pull request by tajila**
Fixes: https://github.com/eclipse-openj9/openj9/issues/19854

### CODE REVIEW

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -234,8 +234,9 @@ static jint runLoadStage (J9JavaVM *vm, IDATA flags);
 #if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
 static void freeClassNativeMemory (J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData);
 #if JAVA_SPEC_VERSION >= 22
-static void vmHookAnonClassesUnload(J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData);
```
Comment (gacholio) [inline runtime/vm/jvminit.c:237]: Why not do this with one hook of the event?

File: runtime/vm/jvminit.c [hunk 2]
```diff
@@ -8205,6 +8207,22 @@ vmHookAnonClassesUnload(J9HookInterface** hook, UDATA eventNum, void* eventData,
...
 #endif /* JAVA_SPEC_VERSION >= 22 */
 
+static void
+vmHookAnonClassesUnloadCleanupJNIIds(J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData)
+{
+	J9VMAnonymousClassesUnloadEvent * unloadedEvent = (J9VMAnonymousClassesUnloadEvent *)eventData;
+	for (J9Class* j9clazz = unloadedEvent->anonymousClassesToUnload; j9clazz; j9clazz = j9clazz->gcLink) {
+		void **jniIDs = j9clazz->jniIDs;
+		if (NULL != jniIDs) {
+			UDATA size = J9VM_SIZEOF_CLASS_JNIID_TABLE(j9clazz->romClass);
```
Comment (gacholio) [inline runtime/vm/jvminit.c:8217]: The macro appears to be size in bytes, not entries. Also, this code should look at the flags to see what behaviour to follow.

File: runtime/vm/jvminit.c [hunk 3]
```diff
@@ -8193,17 +8189,30 @@ freeClassNativeMemory(J9HookInterface** hook, UDATA eventNum, void* eventData, v
...
+	J9JavaVM *vm = vmThread->javaVM;
 	for (J9Class* j9clazz = unloadedEvent->anonymousClassesToUnload; j9clazz; j9clazz = j9clazz->gcLink) {
+		/* AnonClass->classLoader points to the hostclass->classLoader not the anonClassLoader. */
+		if (J9VM_SHOULD_CLEAR_JNIIDS_FOR_ASGCT(vm, j9clazz->classLoader)) {
+			void **jniIDs = j9clazz->jniIDs;
+			if (NULL != jniIDs) {
+				UDATA size = J9VM_SIZEOF_CLASS_JNIID_TABLE(j9clazz->romClass) / sizeof(void*);
+				for (UDATA i = 0; i < size; i++) {
+					J9GenericJNIID *id = (J9GenericJNIID*)(jniIDs[i]);
+					memset(id, 0, sizeof(J9GenericJNIID));
```
Comment (gacholio) [inline runtime/vm/jvminit.c:8207]: The set value should be -1 rather than 0.

File: runtime/vm/jvminit.c [hunk 4]
```diff
@@ -8193,17 +8189,30 @@ freeClassNativeMemory(J9HookInterface** hook, UDATA eventNum, void* eventData, v
...
 {
 	J9VMAnonymousClassesUnloadEvent * unloadedEvent = (J9VMAnonymousClassesUnloadEvent *)eventData;
 	J9VMThread * vmThread = unloadedEvent->currentThread;
+	J9JavaVM *vm = vmThread->javaVM;
 	for (J9Class* j9clazz = unloadedEvent->anonymousClassesToUnload; j9clazz; j9clazz = j9clazz->gcLink) {
+		/* AnonClass->classLoader points to the hostclass->classLoader not the anonClassLoader. */
+		if (J9VM_SHOULD_CLEAR_JNIIDS_FOR_ASGCT(vm, j9clazz->classLoader)) {
+			void **jniIDs = j9clazz->jniIDs;
+			if (NULL != jniIDs) {
+				UDATA size = J9VM_SIZEOF_CLASS_JNIID_TABLE(j9clazz->romClass) / sizeof(void*);
```
Comment (gacholio) [inline runtime/vm/jvminit.c:8204]: This should probably be the opposite - the macro should be entries, and the bytewise use should multiply by the pointer size in ensureJNIIDTable.

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
jenkins test sanity alinux jdk21

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk21

**Comment by gacholio** [comment]
alinux failure is CRIU.

**Pull request merged**